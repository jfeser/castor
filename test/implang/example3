select([p_counter, c_counter],
  depjoin(
    ahashidx(dedup(select([id as p_id], log)),
    alist(select([counter, succ], filter(0.p_id = id && counter < succ, log)),
      atuple([ascalar(0.counter as p_counter), ascalar(0.succ as p_succ)], cross)),
    id_p),
  select([0.p_counter, c_counter],
  filter(c_id = id_c,
    aorderedidx(select([counter], log),
      alist(filter(counter = 0.counter, log),
        atuple([ascalar(0.id as c_id), ascalar(0.counter as c_counter)], cross)),
      0.p_counter, 0.p_succ)))))
