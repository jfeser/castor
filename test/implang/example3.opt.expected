// Locals:
// hoisted0 : int (persists=false)
// hoisted1 : int (persists=false)
// hoisted2 : int (persists=false)
// hoisted3 : int (persists=false)
// hoisted4 : int (persists=false)
// hoisted5 : int (persists=false)
// hoisted6 : int (persists=false)
// hoisted7 : int (persists=false)
// hoisted8 : int (persists=false)
// hoisted9 : int (persists=false)
// hoisted10 : int (persists=false)
// hoisted11 : int (persists=false)
// hoisted12 : int (persists=false)
// hoisted13 : int (persists=false)
// hoisted14 : int (persists=false)
// hoisted15 : int (persists=false)
// hoisted16 : int (persists=false)
// hoisted17 : int (persists=false)
// hoisted18 : int (persists=false)
// hoisted19 : int (persists=false)
// i5 : int (persists=true)
// vstart10 : int (persists=true)
// key17 : tuple[int] (persists=false)
// key11 : tuple[int] (persists=true)
// cstart22 : int (persists=true)
// cstart8 : int (persists=true)
// key24 : tuple[int] (persists=false)
// key2 : tuple[int] (persists=false)
// cstart4 : int (persists=true)
// cstart23 : int (persists=true)
// cstart19 : int (persists=true)
// cstart7 : int (persists=true)
// mid14 : int (persists=true)
// count6 : int (persists=true)
// key15 : tuple[int] (persists=false)
// kstart0 : int (persists=false)
// i20 : int (persists=true)
// hash3 : int (persists=false)
// count21 : int (persists=true)
// vstart1 : int (persists=false)
// idx16 : int (persists=true)
// high13 : int (persists=true)
// key18 : tuple[int] (persists=true)
// low12 : int (persists=true)
// kstart9 : int (persists=true)
fun printer () : void {
    hoisted0 = id_p * 1;
    hoisted1 = buf[2 : 1];
    hoisted2 = hoisted1 - 1;
    hoisted3 = buf[2 : 1];
    hoisted4 = 3 + hoisted3;
    hoisted5 = buf[2 : 1];
    hoisted6 = 3 + hoisted5;
    hoisted7 = 10 / 2;
    hoisted8 = buf[1 : 1];
    hoisted9 = 1 + hoisted8;
    hoisted10 = 10 / 2;
    hoisted11 = buf[1 : 1];
    hoisted12 = 1 + hoisted11;
    hoisted13 = 10 / 2;
    hoisted14 = buf[1 : 1];
    hoisted15 = 1 + hoisted14;
    hoisted16 = buf[1 : 1];
    hoisted17 = 1 + hoisted16;
    hoisted18 = buf[1 : 1];
    hoisted19 = 1 + hoisted18;
    hash3 = hoisted0;
    if (hash3 < 0 || hoisted2 < hash3) {
        
    } else {
         kstart0 = buf[3 + hash3 : 1] + hoisted4;
         key2 = (buf[kstart0 : 1]);
         vstart1 = buf[3 + hash3 : 1] + hoisted6 + 1;
         if (true && key2[0] == id_p) {
             cstart4 = vstart1 + 1 + 1;
             i5 = 0;
             count6 = buf[vstart1 : 1];
             loop (i5 < count6) {
                 cstart7 = cstart4;
                 cstart8 = cstart7 + 1;
                 low12 = 0;
                 high13 = hoisted7;
                 loop (low12 < high13) {
                     mid14 = low12 + high13 / 2;
                     kstart9 = hoisted9 + mid14 * 2;
                     key15 = (buf[kstart9 : 1]);
                     if (not(key15[0] < buf[cstart7 : 1])) {
                         high13 = mid14;
                     } else {
                          low12 = mid14 + 1;
                     }
                 }
                 idx16 = low12;
                 if (idx16 < hoisted10) {
                     kstart9 = hoisted12 + idx16 * 2;
                     key17 = (buf[kstart9 : 1]);
                     key18 = key17;
                     loop (key18[0] < buf[cstart8 : 1] && idx16 < hoisted13) {
                         if (key18[0] < buf[cstart8 : 1] &&
                             not(key18[0] < buf[cstart7 : 1])) {
                             vstart10 =
                             buf[hoisted15 + idx16 * 2 + 1 : 1] + 10 +
                             hoisted17;
                             key11 = key18;
                             cstart19 = vstart10;
                             i20 = 0;
                             count21 = 1;
                             loop (i20 < count21) {
                                 cstart22 = cstart19;
                                 cstart23 = cstart22 + 1;
                                 if (buf[cstart22 : 1] == id_c) {
                                     print(tuple[int, int],
                                     (buf[cstart7 : 1], buf[cstart23 : 1]));
                                 } else {
                                      
                                 }
                                 cstart19 = cstart19 + 2;
                                 i20 = i20 + 1;
                             }
                         } else {
                              
                         }
                         idx16 = idx16 + 1;
                         kstart9 = hoisted19 + idx16 * 2;
                         key24 = (buf[kstart9 : 1]);
                         key18 = key24;
                     }
                 } else {
                      
                 }
                 cstart4 = cstart4 + 2;
                 i5 = i5 + 1;
             }
         } else {
              
         }
    }
}
// Locals:
// hoisted20 : int (persists=false)
// hoisted21 : int (persists=false)
// hoisted22 : int (persists=false)
// hoisted23 : int (persists=false)
// hoisted24 : int (persists=false)
// hoisted25 : int (persists=false)
// hoisted26 : int (persists=false)
// hoisted27 : int (persists=false)
// hoisted28 : int (persists=false)
// hoisted29 : int (persists=false)
// hoisted30 : int (persists=false)
// hoisted31 : int (persists=false)
// hoisted32 : int (persists=false)
// hoisted33 : int (persists=false)
// hoisted34 : int (persists=false)
// hoisted35 : int (persists=false)
// hoisted36 : int (persists=false)
// hoisted37 : int (persists=false)
// hoisted38 : int (persists=false)
// hoisted39 : int (persists=false)
// i5 : int (persists=true)
// vstart10 : int (persists=true)
// key17 : tuple[int] (persists=false)
// key11 : tuple[int] (persists=true)
// cstart22 : int (persists=true)
// cstart8 : int (persists=true)
// key24 : tuple[int] (persists=false)
// key2 : tuple[int] (persists=false)
// cstart4 : int (persists=true)
// cstart23 : int (persists=true)
// cstart19 : int (persists=true)
// cstart7 : int (persists=true)
// mid14 : int (persists=true)
// count6 : int (persists=true)
// key15 : tuple[int] (persists=false)
// kstart0 : int (persists=false)
// i20 : int (persists=true)
// hash3 : int (persists=false)
// count21 : int (persists=true)
// vstart1 : int (persists=false)
// idx16 : int (persists=true)
// high13 : int (persists=true)
// key18 : tuple[int] (persists=true)
// low12 : int (persists=true)
// kstart9 : int (persists=true)
fun consumer () : void {
    hoisted20 = id_p * 1;
    hoisted21 = buf[2 : 1];
    hoisted22 = hoisted21 - 1;
    hoisted23 = buf[2 : 1];
    hoisted24 = 3 + hoisted23;
    hoisted25 = buf[2 : 1];
    hoisted26 = 3 + hoisted25;
    hoisted27 = 10 / 2;
    hoisted28 = buf[1 : 1];
    hoisted29 = 1 + hoisted28;
    hoisted30 = 10 / 2;
    hoisted31 = buf[1 : 1];
    hoisted32 = 1 + hoisted31;
    hoisted33 = 10 / 2;
    hoisted34 = buf[1 : 1];
    hoisted35 = 1 + hoisted34;
    hoisted36 = buf[1 : 1];
    hoisted37 = 1 + hoisted36;
    hoisted38 = buf[1 : 1];
    hoisted39 = 1 + hoisted38;
    hash3 = hoisted20;
    if (hash3 < 0 || hoisted22 < hash3) {
        
    } else {
         kstart0 = buf[3 + hash3 : 1] + hoisted24;
         key2 = (buf[kstart0 : 1]);
         vstart1 = buf[3 + hash3 : 1] + hoisted26 + 1;
         if (true && key2[0] == id_p) {
             cstart4 = vstart1 + 1 + 1;
             i5 = 0;
             count6 = buf[vstart1 : 1];
             loop (i5 < count6) {
                 cstart7 = cstart4;
                 cstart8 = cstart7 + 1;
                 low12 = 0;
                 high13 = hoisted27;
                 loop (low12 < high13) {
                     mid14 = low12 + high13 / 2;
                     kstart9 = hoisted29 + mid14 * 2;
                     key15 = (buf[kstart9 : 1]);
                     if (not(key15[0] < buf[cstart7 : 1])) {
                         high13 = mid14;
                     } else {
                          low12 = mid14 + 1;
                     }
                 }
                 idx16 = low12;
                 if (idx16 < hoisted30) {
                     kstart9 = hoisted32 + idx16 * 2;
                     key17 = (buf[kstart9 : 1]);
                     key18 = key17;
                     loop (key18[0] < buf[cstart8 : 1] && idx16 < hoisted33) {
                         if (key18[0] < buf[cstart8 : 1] &&
                             not(key18[0] < buf[cstart7 : 1])) {
                             vstart10 =
                             buf[hoisted35 + idx16 * 2 + 1 : 1] + 10 +
                             hoisted37;
                             key11 = key18;
                             cstart19 = vstart10;
                             i20 = 0;
                             count21 = 1;
                             loop (i20 < count21) {
                                 cstart22 = cstart19;
                                 cstart23 = cstart22 + 1;
                                 if (buf[cstart22 : 1] == id_c) {
                                     consume(tuple[int, int],
                                     (buf[cstart7 : 1], buf[cstart23 : 1]));
                                 } else {
                                      
                                 }
                                 cstart19 = cstart19 + 2;
                                 i20 = i20 + 1;
                             }
                         } else {
                              
                         }
                         idx16 = idx16 + 1;
                         kstart9 = hoisted39 + idx16 * 2;
                         key24 = (buf[kstart9 : 1]);
                         key18 = key24;
                     }
                 } else {
                      
                 }
                 cstart4 = cstart4 + 2;
                 i5 = i5 + 1;
             }
         } else {
              
         }
    }
}
